### PR TITLE
Pull request of the testing branch created for submittion for code review 

### PR DESCRIPTION
**Pull request by aulbytj**
Testing your Code with RSpec.
In this pull request, we created various tests for the Enumerable methods to test that the methods operate as expected.

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    end
+
+    it 'should returns key, value pair' do
+      expect({ name: 'jay', age: 34 }.my_each.to_h).to(eq({ name: 'jay', age: 34 }.each.to_h))
+    end
+
+    it 'returns elements in the range' do
+      expect(%i[0..10].my_each.to_a).to(eq(%i[0..10].each.to_a))
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: I can see a lot of repetitive arrays, hash, range. Which is not a good sign of clean code. First of all, you can assign them in a variable. then you can them whenever needed. In Rspec we assign variables using `let`.And my_each is not checking for some edge case. I tried my best to organize your test in this format. You can either replace your one with it or you can generate a new case using the following format.
```ruby
describe '#my_each' do
    let(:array) { [1, 2, 4, 5, 6] }
    let(:range) {Range.new(0, 5)}
    let(:hash) { {name: 'jay', age: 34} }
    it 'returns enumerator' do
      expect(array.my_each).to be_an(Enumerator)
    end

    it 'calls the block for each of the item in the array' do
      each_output = 0
      block = proc { |num| each_output += num }
      array.my_each(&block)
      my_each_output = each_output.dup
      each_output = 0
      array.each(&block)
      expect(my_each_output).to eql(each_output)
    end

    it 'calls the block for each of the item in the hash' do
      each_output = ''
      block = proc { |key, val| each_output += "#{key} #{val}" }
      hash.my_each(&block)
      my_each_output = each_output.dup
      each_output = ''
      hash.each(&block)
      expect(my_each_output).to eql(each_output)
    end

    it 'calls the block for each of the item in the array' do
      each_output = 0
      block = proc { |num| each_output += num }
      range.my_each(&block)
      my_each_output = each_output.dup
      each_output = 0
      range.each(&block)
      expect(my_each_output).to eql(each_output)
    end
end
```

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  describe '#my_each_with_index' do
+    it 'returns enumerator' do
+      expect([1, 2, 'ten', 3.5, %w[hello world!]].my_each).to(be_an(Enumerator))
+    end
+
+    it 'should return the indexes' do
+      array = []
+      %w[cat dog fish].my_each_with_index { |_word, idx| array[idx] = idx }
+      expect(array).to(eq([0, 1, 2]))
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You can compare it with each_with_index by following the example I have given above.

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  end
+
+  describe '#my_select' do
+    it 'should return the even values' do
+      expect([1, 2, 3, 4, 5, 6, 7, 8, 9].select { |n| n % 3 == 0 }).to(eq([3, 6, 9]))
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You have used the actual select method. I cannot see the use of my_select. Please fix it.

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  end
+
+  describe '#my_any?' do
+    it 'should return true' do
+      expect([nil, true, 99].any?).not_to(eq(false))
+    end
+
+    it 'should return false' do
+      expect(%w[ant bear cat].any?(/d/)).to(eq(false))
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: - You have called any?, But you were suppose to call my_any?
- What about the class? As a parameter, It can accept a class name.
- my_any? can accept a value as a parameter. I can not see any tests related to that.

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  describe '#my_all?' do
+    it 'should return true' do
+      expect([].all?).to(eq(true))
+    end
+
+    it 'should return false' do
+      expect(%w[ant bear cat].all? { |word| word.length >= 4 }).to(eql(false))
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Are you sure you are testing my_all?. I can see all? in here inside the expect block. To make sure you are supposed to tests your enumerable methods.
- Add a test related to a class as a parameter.
- Add tests related to value as a parameter.
- Add tests related to regex as a parameter.

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+  describe '#my_count' do
+    it 'should return number of items in array' do
+      expect([61, 51, 41, 31, 21, 11].my_count).to(eq(6))
+    end
+
+    it 'should return zero' do
+      expect([64, 52, 41, 31, 22, 11].my_count { |n| n % 3 == 0 }).to(eq(0))
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Let's assign the array inside a variable by using let.
[Optional]
You can compare the test with the actual count method.

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  end
+
+  describe '#my_map' do
+    it 'should return new array with [1, 4, 9, 16]' do
+      expect((1..4).map { |i| i * i }).to(eq([1, 4, 9, 16]))
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You are testing map instead of my_map. Please fix the issue

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+  describe '#none' do
+    it 'should return true' do
+      expect([nil, false].none?).to(eq(true))
+    end
+
+    it 'shoudl not return false' do
+      expect(%w[ant bear cat].none? { |word| word.length == 5 }).to_not(eq(false))
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Describe should indicate my_none. And test should be related to my_none also. But I can see none?, why is that?
- Add test to my_none when a regex is given as a parameter.
- Add test to my_none when a class is given as a parameter
- Add test to my_none when a value is given as a parameter.

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      longest =
+        %w[cat sheep bear].reduce do |memo, word|
+          memo.length > word.length ? memo : word
+        end
+      expect(longest).to(eq('sheep'))
+    end
+
+    it 'should return 151200' do
+      expect((5..10).reduce(1, :*)).to(eq(151_200))
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: I don't think that you are checking the my_inject. It seems that you are testing builtin reduce method of ruby. So please update the test and make sure that it is related to my_inject.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @aulbytj and @appiersign :wave: 
Great initiative, But to achieve the ultimate goal you need to cover some cases. So let's get to them.
___
### :spades: Mandatory Changes :spades: 
___
1. I cannot see any info in your readme related to testing. Can you add the info, like how to execute the test?
2. It seems that almost all of your methods contain some failing edges. Check the comments related to each testing block. I have added a sample format as a template. You can check it.
___
#### Please fix the above-mentioned issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello Team!
###  You have done really well and for also making the changes from the previous reviews :clap: However, you have a few issues to address
- Solve the issues raised by rubocop
### Make the changes and resubmit for another code review. Good luck.
- Happy coding!

By Raymond
slack @iamraymond

**Review by AdnanAfsari — APPROVED** [review]
**Status: Approved**

Hello @aulbytj @appiersign ✋ 

Nice work, It's a 👍 from me.

Happy Coding
Adnan Afsari
